### PR TITLE
[TESTERS NEEDED] rsx: Handle tiled memory

### PR DESCRIPTION
**Pull request by kd-11**
Addresses https://github.com/RPCS3/rpcs3/issues/13680

Fixes https://github.com/RPCS3/rpcs3/issues/6399
Fixes https://github.com/RPCS3/rpcs3/issues/6609 (partially, that game has another bug)
Fixes https://github.com/RPCS3/rpcs3/issues/12756

Also fixes many AAA games when run without patches + WCB+RCB because they often do SPU MLAA. This set fixes memory access and removes the honeycomb stippling pattern observed previously.

### CODE REVIEW

**Comment by MarioSonic2987** [comment]
Fixes the honeycomb stippling glitch in Split/Second.

**Before**:
image

**After**:
image

**Comment by TigiHof** [comment]
Fixes the honeycomb stippling glitch in Star Wars: The Force Unleashed Demo (NPEB90094).

Before:
NPEB90094_screenshot_2023_09_15_03_37_08

After:
NPEB90094_screenshot_2023_09_15_03_26_40

However notice the graphical artifact in the bottom right corner.

Log File:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/12614781/RPCS3.log.gz)

RSX Capture:

File: rpcs3/emucore.vcxproj.filters [hunk 1]
```diff
@@ -2332,7 +2335,10 @@
     <ClInclude Include="Emu\NP\upnp_config.h">
       <Filter>Emu\NP</Filter>
     </ClInclude>
-    <ClInclude Include="io_buffer.h">
+    <ClInclude Include="Emu\RSX\Common\io_buffer.h">
+      <Filter>Emu\GPU\RSX\Common</Filter>
+    </ClInclude>
+	<ClInclude Include="Emu\RSX\Common\tiled_dma_copy.hpp">
```
Comment (Megamouse) [inline rpcs3/emucore.vcxproj.filters:2341]:
```suggestion
    <ClInclude Include="Emu\RSX\Common\tiled_dma_copy.hpp">
```

**Comment by Ordinary205** [comment]
Partially fixes NFS The Run when enabling WCB+RCB.

Master:
Need for Speed The Run Master

PR:
Need for Speed The Run PR

Also the RCB seems to have fixed the missing lighting, although they still appear incorrect.

**Comment by cas-1092** [comment]
Partially fixes artifacting in Blur with RCB.
PR
image
Master
image

**Comment by Ordinary205** [comment]
Fixes disney games such as Cars 3: Driven to Win.

Master:
Cars 3 3

PR:
Cars 3 fix
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/12619911/RPCS3.log.gz)

**Comment by Ordinary205** [comment]
This also fixes incorrect car reflections on NFS Most Wanted and Hot Pursuit.

Master:
NFS MW 1

PR:
Need for speed most wanted

Apologizes if I keep posting bug fix messages.

**Comment by Linear524** [comment]
Holy s#1D !!! I've been waiting many years for this ! :) Thank you KD-11 ! =D
MLAA is finaly fixed for many titles ! And it is almost perfectly fixed, just some minor issues with some horisontal line (looks like overflow somewhere)

**Comment by Linear524** [comment]
Uncharted 2 is fully playable now and no patches needed :)
Native 720p mode with FSR upscale and with RCB+WCB+RDB+WDB on.

**Comment by Linear524** [comment]
Uncharted 3 is 99% playable:



And here is that 1% issue with horizontal artifacts:

**Comment by Linear524** [comment]
Last of Us is 90% playable now without any patches, just Native 720p mode with FSR upscale and with RCB+WCB+RDB+WDB on:

And the other 10% issues =  is missing vlolumetric lighting via SPUs, missing smoke and fog via SPUs and those horizontal lines artifacts when there is very bright object in rendered frame (overflow in color buffers?) :

**Comment by elad335** [comment]
Please test regressions with savestates with this.

**Comment by Ordinary205** [comment]
Mountain textures now renders correctly on NFS Rivals when using WCB+RCB.
Need for Speed Rivals 1
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/12622512/RPCS3.log.gz)

**Comment by YuriNator557** [comment]
Wasn't a reported issue on GitHub, but graphics in Sniper Elite III are vastly improved

Master (as of 15468)
image
image

PR
image
image

**Comment by scorpion81** [comment]
I know this is still a work in progress PR, but I wanted to list some remaining issues after testing Split / Second so far.
Nevertheless this is already a great improvement :) 

1) In the game menu, there are still missing explosion and smoke effects.
2) In the car selection menu, there are ghosting issues when the car models are being rotated or moved.
3) Without WCB+RCB, the skybox smearing / ghosting issue still is present, and some environment lighting seems to miss.
4) With WCB + RCB, the skybox smearing and environment lighting is fixed, but ingame fire and smoke effects are missing.
Side note, without WCB + RCB performance seems also be better by 5 - 10 fps.

I have also tested with WDB + RDB but both didnt seem to have an effect.

**Comment by Asinin3** [comment]
@Linear524 Would be good if you could show a comparison of what was actually fixed or write it. And less images. I spent awhile comparing and cant see whats changed compared to master.

@cas-1092 I spent almost an hour with Blur and dont get that issue on master. RCB actually made it worse for me and I was 60fps locked even with accurate xfloat at 400% res scale. Graphics seemed fine on master? 

Also having issues getting ingame on Force Unleashed and NFS Rivals, if someone can msg me on our discord server and provide footage that would be helpful.

**Comment by Ordinary205** [comment]
Here's a full comparison that I've tested on five games.
https://youtu.be/VMPkN_DRfF8
Sorry if this was recorded on 1080p.

**Comment by Ordinary205** [comment]
I forgot to mention that this causes a MLAA regression on few frostbite games. Which appears yellow for some reason.
Need for Speed Rivals regression
Battlefield 3

But if I disable the MLAA on NFS Rivals patches, the yellow corrupted screen is gone.
Need for Speed Rivals without MLAA

**Comment by Linear524** [comment]
@Asinin3 
Okay. I will try to describe = UC2 was messed up before, when you move camera in front of the flashlight - there was broken MLAA artifacts, and very bright areas was giving the same MLAA artifacts. Now it is fixed with this PR. Even main menu artifacts is fixed by the way. And working MLAA gives us more antialiased edges on the most of rendered geometry. 
(NaughtyDog doesn't use MLAA on every geometry layer, just partially on more significant objects)

**Comment by elad335** [comment]
An image is worth a thousand words.

**Comment by Ordinary205** [comment]
Here's a comparison for Star Wars: The Force Unleashed.
https://drive.google.com/file/d/164PGETpFuLgc7V53zq0qVAjxn9q7vVzZ/view?usp=drive_link

**Comment by jenci8888** [comment]
NFS Hot Pursuit 2010 is fixed reflection issue (glossy paint and glass)

Master:
image
PR:
image

**Comment by TryTwo** [comment]
Fixes The Witch and the Hundred Knight honeycomb stippling #7041

BLUS30964_screenshot_2023_09_17_10_21_13

**Comment by legend800** [comment]
@TigiHof That gfx artifact can be fixed with SPU Float Accurate.:
2023-09-17 14_13_10-FPS_ 29 96 _ Vulkan _ 0 0 29-70acf961 Alpha _ rsx-tiling _ Star Wars_ The Force

So settings for this game are: WCB + RCB(general gfx), Accurate SPU DMA(boot fix), SPU Float Accurate(corrupted border gfx)

The game appears to be perfect other than an issue where all the main/pause menu text gets corrupted on the 2nd load. Tried every setting possible but couldn't find fix. Present on Master as well. 
2023-09-17 14_16_31-FPS_ 29 35 _ Vulkan _ 0 0 29-70acf961 Alpha _ rsx-tiling _ Star Wars_ The Force

2023-09-17 14_20_02-FPS_ 30 00 _ Vulkan _ 0 0 29-70acf961 Alpha _ rsx-tiling _ Star Wars_ The Force

**Comment by TigiHof** [comment]
> @TigiHof That gfx artifact can be fixed with SPU Float Accurate.

**Comment by kd-11** [comment]
> > @TigiHof That gfx artifact can be fixed with SPU Float Accurate.
> 
> Unfortunately that does not work for me. The artifact is still present. But thanks, anyway!

I cannot reproduce the artifact. Maybe it's caused by upscaling in this case? Missing pixels in the bottom-right is a known gotcha with the tiling setup because the hardware overscans the memory by 48px at 720p res. However, that is already accounted for.

**Comment by Darkhost1999** [comment]
image
Turing gfx gtx 1660 super with default resolution
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/12648227/RPCS3.log)
Accurate rsx reservation access resolving freezes for me.
Accurate xfloat doesn't change the image

**Comment by TigiHof** [comment]
> I cannot reproduce the artifact. Maybe it's caused by upscaling in this case? Missing pixels in the bottom-right is a known gotcha with the tiling setup because the hardware overscans the memory by 48px at 720p res. However, that is already accounted for.

No, no upscaling involved. Here is a RSX capture:

**Comment by kd-11** [comment]
I think I'll just try and figure out why rivals broke and commit this. There is a lot of work left to do afterward such as OpenGL support which is completely missing at the moment.

**Comment by Augusto7743** [comment]
Possibly fix Motorstorm Pacific Rift too ...
Where have an link to download that PR#14647 Linux build for test ?

**Comment by Asinin3** [comment]
What do you mean by Pacific Rift, it doesn't have this issue? All it has is the bloom post-processing artifact when upscaling but thats not related to this issue

**Comment by Ordinary205** [comment]
I understand that this PR is still work in progress or might not fix this issue, but the other graphical bug for vehicle artifact reflections on NFS The Run seems to behave differently.
Before:
NFS The Run before

After: As you can see, the artifacts appears outside of the vehicle.
NFS The Run after

**Comment by Augusto7743** [comment]
@Asinin3

NFS Hot Pursuit wrong colors in cars was fixed.
Motorstorm Pacific Rift need WCB to fix colors ... perhaps that PR fix it

**Comment by Darkhost1999** [comment]
If Write color buffers was needed before then Write color buffers will be needed after.

> Why?

Because simply stated all the buffers are true to how the PS3 works. Making them more accurate.

> So then why aren't all the buffers enabled by default?

Accuracy settings are not as fast as speed hacks. Your computer is known to be well below the recommend specs. If we enabled these by default you'd be forced to acquire a new pc to play your games and supposedly even next generation PCs have performance issues in some cases.
Nevertheless https://rpcs3.net/roadmap
Read buffers will be enabled by default as apart of short term goals if the goals can be achieved.

**Comment by Danke-Boi** [comment]
> "Your computer is known to be well below the recommend specs" Ironically run all my PS3 console games ( GT5P, RE5, TMNT, Spelunker and some others) with good speed.

He specifically said that the reason why you run all of the games you do while having hardware that is below recommended specs is because they don't have Write Color Buffers enabled. The more accurate the emulation, the harder it is to emulate. Those games are using hacks to make it possible for your pc to run the games well. He's also letting you know that since your hardware is below recommended specs that updates in the future will affect how well you can run your games.

**Comment by Augusto7743** [comment]
@Danke-Boi
I had understood ... however not will have an option to enable and disable WCB+WDB in future builds ?

Not any Linux build in link below for test
https://cirrus-ci.com/github/RPCS3/rpcs3

**Comment by Linear524** [comment]
RCB+WCB+RDB+RDB should always be enabled technically . Because it is what real PS3 does ! And because a lot of graphical functions will be missed without this.

**Comment by Augusto7743** [comment]
Thanks for sharing that information.
If will be done an Linux build I will test it.

**Comment by kd-11** [comment]
I think temporarily it makes sense to make this an advanced option until the rivals issue is properly fixed. Unfortunately there are too many paths that write to memory at the moment and I don't want to increase complexity of the PR. I did find one bug however, I'll push a fix before merge.

**Comment by Triticum0** [comment]
Improves The Chronicles of Riddick Assault on Dark Athena. still not perfect in motion
Before 
NPEB90133_screenshot_2023_01_15_00_36_50
NPEB90133_screenshot_2023_01_15_00_37_19

After
NPEB90133_screenshot_2023_09_21_17_46_56
NPEB90133_screenshot_2023_09_21_17_46_53

**Comment by Triticum0** [comment]
@kd-11 I also found issues with mercury hg and Retro/Grade having duplication of menu items when in menus and ingame also doesn't fix mercury hg at all.

**Comment by Triticum0** [comment]
mercury hg
Master
NPUB30419_screenshot_2023_09_21_17_56_41

PR
NPUB30419_screenshot_2023_09_21_17_59_53

Retro/Grade
Master
NPUB30280_screenshot_2023_09_21_17_57_21
NPUB30280_screenshot_2023_09_21_17_58_36

PR
NPUB30280_screenshot_2023_09_21_18_00_55
NPUB30280_screenshot_2023_09_21_18_02_33

**Comment by Triticum0** [comment]
[mercury hg.log.gz](https://github.com/RPCS3/rpcs3/files/12689384/RPCS3.log.gz)

**Comment by Triticum0** [comment]
Regression in Army of TWO The Devil's Cartel
Master
NPEB90470_screenshot_2023_09_21_18_28_17

PR
NPEB90470_screenshot_2023_09_21_18_30_01

**Comment by Ordinary205** [comment]
> > Regression in Army of TWO The Devil's Cartel Master 
> 
> 
> 
> This happens in every frostbite 2.0+ game

Not all frostbite games, NFS The Run is unaffected by this issue, I think.

**Comment by Asinin3** [comment]
> mercury hg Master NPUB30419_screenshot_2023_09_21_17_56_41
> 
> PR NPUB30419_screenshot_2023_09_21_17_59_53
> 
> Retro/Grade Master NPUB30280_screenshot_2023_09_21_17_57_21 NPUB30280_screenshot_2023_09_21_17_58_36
> 
> PR NPUB30280_screenshot_2023_09_21_18_00_55 NPUB30280_screenshot_2023_09_21_18_02_33

The mercury issue you showed on PR was present in master  I think you just have run to run variance, i remember seeing that issue 6months ago as well as two years ago.

**Comment by kd-11** [comment]
NFS rivals is now fixed.

**Comment by Ordinary205** [comment]
@kd-11 The newest commit causes yellow screen regression on NFS The Run.
image
Sorry if this photo wasn’t taken on the PC.

**Comment by Ordinary205** [comment]
NFS Rivals is still not fixed.

**Comment by Augusto7743** [comment]
not any linux build for test ?
only win10 and freebsd are available.

**Comment by cipherxof** [comment]
> not any linux build for test ? only win10 and freebsd are available.

https://dev.azure.com/nekotekina/nekotekina/_build/results?buildId=8589&view=artifacts&pathAsName=false&type=publishedArtifacts

image

image

**Comment by kd-11** [comment]
> @kd-11 The newest commit causes yellow screen regression on NFS The Run. image Sorry if this photo wasn’t taken on the PC.

Please attach the log file. Also check if enabling all paths (WCB,WDB,RCB,RDB) changes anything.

**Comment by kd-11** [comment]
> NFS Rivals is still not fixed.

I verified it myself. I think what is left is a settings issue.

**Comment by Ordinary205** [comment]
Enabling WCB+RCB with WDB+RDB doesn't solve any issues for Rivals and The Run.
Still happens
[NFSTheRun.gz](https://github.com/RPCS3/rpcs3/files/12705636/NFSTheRun.gz)

Still happens aswell
[NFSRivals.gz](https://github.com/RPCS3/rpcs3/files/12705644/NFSRivals.gz)

**Comment by kd-11** [comment]
For whatever reason, it works fine for me:
image

I'll have to think of a way to repro this.

**Comment by kd-11** [comment]
Nvm, it is a settings issue on my side + some local debug code that affected the results.

**Comment by MastaG** [comment]
Perhaps not the right place to ask this. But you mention there is still a lot of work to do for OpenGL... So what advantages does this API offer over Vulkan? Isn't it better to drop it and use Vulkan exclusively?

**Comment by Darkhost1999** [comment]
https://github.com/RPCS3/rpcs3/issues/13929#issuecomment-1567657740

OpenGL tends to have better synchronization with CELL and is generally more accurate.

**Comment by Augusto7743** [comment]
Why OpenGL have better synchronization with Cell than using Vulkan ?
Simply because is more time to create shaders so the emulated Cell have enough time to do some tasks ?

**Comment by Linear524** [comment]
Because real PS3 uses modified version of OpenGL and it is important to replicate this as close as possible for future accuracy fixes !

**Comment by Augusto7743** [comment]
Vulkan not is more or less opengl "5" ? Vulkan not use code from Opengl source ?
Well ... thus using Opengl is more accurate, but is more slow to create and load shaders and performance is more slow than Vulkan too.
More powerful is the GPU more faster will be Vulkan than Opengl.

**Comment by ArchSonix** [comment]
> Vulkan not is more or less opengl "5" ? Vulkan not use code from Opengl source ? Well ... thus using Opengl is more accurate, but is more slow to create and load shaders and performance is more slow than Vulkan too. More powerful is the GPU more faster will be Vulkan than Opengl.

Vulkan is a completely different Graphics API based on AMDs MANTLE, not in any way related to OpenGL.
Both APIs are supported equally. Vulkan is sometimes better for performance and modern hardware features. OpenGL is supported, as in some ways it is closer to the way a real PS3 works.

**This MR is not the correct place to discuss this! If you have questions or want to talk about it, do so in Discord.**

**Comment by kd-11** [comment]
After thinking about it and seeing all the problems I think an option makes more sense than trying to fix everything in one PR.

**Comment by kd-11** [comment]
If any problems are observed, please report them at https://github.com/RPCS3/rpcs3/issues/14682 so I can investigate.

**Comment by alanjsebas** [comment]
Can someone share their build of the Windows version with the fix installed?

**Comment by Jonathan44062** [comment]
@alanjsebas The PR has already been merged in the master, update the emulator and activate this new option in Advanced tab
image

**Comment by alanjsebas** [comment]
Somehow I didn't notice that.

Incredibly sorry, and thanks.

**Comment by Volga342** [comment]
Screenshot (162)
Screenshot (163)
I'm still having this problem. RPCS3 says it's up to date, I've set the config to the recommend settings and it's still happening

**Pull request merged**